archive/imo/imo1969_q1.lean

```diff
@@ -3,67 +3,71 @@ Copyright (c) 2020 Kevin Lacker. All rights reserved.
-which nlinarith can solve since they are expressed as a sum of squares.
+which `nlinarith` can solve since they are each expressed as a sum of squares.
 -/
-lemma left_factor_large (m n : ℤ) (h: 1 < m) : 1 < ((m - n)^2 + m^2) := by nlinarith
-lemma right_factor_large (m n : ℤ) (h: 1 < m) : 1 < ((m + n)^2 + m^2) := by nlinarith
+lemma left_factor_large {m : ℤ} (n : ℤ) (h : 1 < m) : 1 < ((n - m)^2 + m^2) := by nlinarith
+lemma right_factor_large {m : ℤ} (n : ℤ) (h : 1 < m) : 1 < ((n + m)^2 + m^2) := by nlinarith
 
 /-!
 The factorization is over the integers, but we need the nonprimality over the natural numbers.
 -/
-
-lemma int_large (a : ℤ) (h : 1 < a) : 1 < a.nat_abs :=
+lemma int_large {a : ℤ} (h : 1 < a) : 1 < a.nat_abs :=
 by exact_mod_cast lt_of_lt_of_le h le_nat_abs
 
-lemma int_not_prime (a b : ℤ) (c : ℕ) (h1 : 1 < a) (h2 : 1 < b) (h3 : a*b = ↑c) : ¬ prime c :=
-have h4 : (a*b).nat_abs = a.nat_abs * b.nat_abs, from nat_abs_mul a b,
-have h5 : a.nat_abs * b.nat_abs = c, by finish,
-tactic.norm_num.not_prime_helper a.nat_abs b.nat_abs c h5 (int_large a h1) (int_large b h2)
+lemma not_prime_of_int_mul' {a b : ℤ} {c : ℕ}
+  (ha : 1 < a) (hb : 1 < b) (hc : a*b = (c : ℤ)) : ¬ prime c :=
+not_prime_of_int_mul (int_large ha) (int_large hb) hc
 
-lemma polynomial_not_prime (m n : ℕ) (h1 : 1 < m) : ¬ prime (n^4 + 4*m^4) :=
-have h2 : 1 < of_nat m, from coe_nat_lt.mpr h1,
+/-- Every natural number of the form `n^4 + 4*m^4` is not prime. -/
+lemma polynomial_not_prime {m : ℕ} (h1 : 1 < m) (n : ℕ) : ¬ prime (n^4 + 4*m^4) :=
+have h2 : 1 < (m : ℤ), from coe_nat_lt.mpr h1,
 begin
-  refine int_not_prime _ _ _ (left_factor_large ↑m ↑n h2) (right_factor_large ↑m ↑n h2) _,
-  rw factorization,
-  norm_cast
+  refine not_prime_of_int_mul' (left_factor_large (n : ℤ) h2) (right_factor_large (n : ℤ) h2) _,
+  exact_mod_cast factorization
 end
 
-/-!
-Now we just need to show this works for an arbitrarily large $a$, to prove there are
-infinitely many of them.
-$a = 4*(2+b)^4$ should do. So $m = 2+b$.
+/--
+We define $a(b) := 4*(2+b)^4$, so that we can take $m = 2+b$ in `polynomial_not_prime`.
 -/
+def a (b : ℕ) : ℕ := 4*(2+b)^4
 
-theorem imo1969_q1 : ∀ b : ℕ, ∃ a : ℕ, a ≥ b ∧ ∀ n : ℕ, ¬ prime (n^4 + a) :=
-assume b,
-have h1 : 1 < 2+b, by linarith,
-have b^2 ≥ b, by nlinarith,
-have h2 : 4*(2+b)^4 ≥ b, by nlinarith,
-begin
-  use [4*(2+b)^4, h2],
-  assume n,
-  exact polynomial_not_prime (2+b) n h1
-end
+lemma a_good (b : ℕ) : a b ∈ good_nats :=
+polynomial_not_prime (show 1 < 2+b, by linarith)
+
+/-- `a` is a strictly monotone function; this is easily proven by chaining together lemmas
+in the `strict_mono` namespace. -/
+lemma a_strict_mono : strict_mono a :=
+((strict_mono_id.const_add 2).nat_pow (dec_trivial : 0 < 4)).const_mul (dec_trivial : 0 < 4)
+
+/-- We conclude by using the fact that `a` is an injective function from the natural numbers to
+the set `good_nats`. -/
+theorem imo1969_q1 : set.infinite good_nats :=
```

Comment:
I think with these kind of stand-alone formalizations it is nice if the final result is as readable as possible, so I suggest to unfold the definition of `good_nats` here.
```suggestion
theorem imo1969_q1 : set.infinite {a : ℕ | ∀ n : ℕ, ¬ prime (n^4 + a)} :=
```